### PR TITLE
[KVStore] DGL Distributed KVStore

### PR DESCRIPTION
**Pull request by aksnzhy**
### Basic Components

DGL kvstore contains two components: kvserver and kvclient. We can create kvserver and kvclient by using the following APIs:

kvserver:

```python
   server = dgl.contrib.KVServer(
        server_id=id, 
        client_namebook=client_namebook, 
        server_addr='127.0.0.1:50051')
```

kvclient:

```python
   client = dgl.contrib.KVClient(
        client_id=id, 
        server_namebook=server_namebook, 
        client_addr='127.0.0.1:50052')
```

We can start multiple kvserver and kvclient on different machines, and each server/client will have its own unique ID.

 The basic data structure transferred between kvserver and kvclient is tensor (mx.ndarray or torch.tensor). We can initialize tensor data on kvserver by using the following API:

```python
server.add_data('node_embed', torch.zeros([100, 3]))
server.add_data('relation_embed', torch.zeros([30, 3]))
```

Unlike MXNet kvstore, DGL kvstore gives each type of data a name for supporting multiple-type data, which is commonly used in knowledge graph and GNN tasks. 

After initialization, we can start kvserver like this:

```python
server.start()
```

To maximize the throughput of kvserver, we use multi-threading and message queue to receive data from different clients. After starting kvserver, kvclient can connect to kvserver:

```python
client.connect()
```

### PUSH / PULL

After connecting to kvserver, kvclient can push/pull data to/from kvserver. 

For example:

```python
   id = torch.tensor([15, 33, 57])
   data = torch.tensor([[0., 0., 0., ], [1., 1., 1.], [2., 2., 2.]])

   client.push('node_embed', id, data)
   new_data = client.pull('node_embed', id)
```

In this example, we first push 3 node-embeddings with the ID [15, 33, 57] to kvserver, and then pull the updated embedding from it. The `new_data`  will be:

```python
tensor([[0., 0., 0., ], [1., 1., 1.], [2., 2., 2.]]
```

### Computation on KVStore

On default, the kvserver will do `sum` operation when receiving a push message, and it will do `index_copy` operation when receiving a pull message. Developers can re-implement these two default operations by override _push_handler and _pull_handler on kvserver. 

```python
   class MyKVServer(KVServer):

      def _push_handler(slef, name, ID, data):
          #... User-defined code ... 

      def _pull_handler(slef, name, ID):
          #... User-defined code ... 
```

### Data Partition Scheme

DGL kvstore uses a very simple range-partition scheme to partition data into different server nodes. For example, if the total embedding count is 200 and we have two kvserver, the data [0~99] will be stored in kvserver_0, and data [100~199] will be stored in kvserver_1.  The push() and pull() API will mapping data to corresponded server node automatically.

### Networking

The underlying networking infrastructure used by DGL is a lightweight send/recv multi-threading communication wrapper for tcp socket (https://github.com/dmlc/dgl/tree/master/src/graph/network). 

Using this send/recv interface, we can easily extend kvstore to the other networking hardware environment and libraries such as mpi and zmq.

### CODE REVIEW

**Comment by zheng-da** [comment]
Can you add `init_data` in the client side?

**Comment by aksnzhy** [comment]
> Can you add `init_data` in the client side?

Sure, I'm working on this.

**Comment by futurely** [comment]
I would like to introduce some alternative design choices of another gnn framework Euler.

**Server**

The graph service [use the server](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/service/graph_service.h#L44) generated by the [grpc framework](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/service/graph_service.h#L26) to [handle](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/service/graph_service.cc#L63) client requests.

User can [start or start_and_wait](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/python/service.py) the [service](https://github.com/alibaba/euler/search?q=StartService&unscoped_q=StartService).



**Client**

The [client ](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/client/rpc_client.h) is [only used in C++](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/client/remote_graph_shard.h#L117). There is no python version client. So users are unaware of them.

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/client/rpc_manager.h

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/client/grpc_manager.h

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/client/grpc_channel.h

**Cluster membership**

ZooKeeper is used to register and monitor servers presence in the cluster.

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/common/server_register.h

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/common/zk_server_register.cc

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/common/server_monitor.cc

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/common/zk_server_monitor.cc

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/common/zk_service.cc

**PUSH / PULL**

The RPC messages are [defined with protocol buffer](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/proto/graph_service.proto) and [wrapped as various CallData](https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/service/call_data.h).

https://github.com/alibaba/euler/blob/69651bfb7ffd44de1b83b56e199f348665158f50/euler/service/graph_service.cc#L63

**Computation on KVStore**

**Data Partition Scheme**

**Networking**

**Comment by yzh119** [comment]
@futurely , thanks for your suggestions, looks like you are quite familiar with Euler, so could you please create issues to discuss the designs rather than comment on many closed PRs?
@aksnzhy has compared custom networking and grpc, and the speed of custom networking is significantly faster.
We welcome your contributions if you have better designs.

File: src/graph/network/serialize.cc [hunk 1]
```diff
@@ -161,5 +161,127 @@ void DeserializeSampledSubgraph(char* data,
...
+  int64_t id_size = 0;
+  int64_t data_size = 0;
+  int64_t id_shape = 0;
+  int64_t data_shape_0 = 0;
+  int64_t data_shape_1 = 0;
+  total_size += sizeof(msg_type);
+  total_size += sizeof(rank);
+  if (name != nullptr) {
+    total_size += sizeof(name->length());
+    total_size += name->length();
```
Comment (classicsong) [inline src/graph/network/serialize.cc]: I think we can add an alignment for string name. which may help speedup memcpy

**Pull request closed without merging**